src/dune/build_system.ml

```diff
@@ -1439,11 +1439,14 @@ end = struct
         match from_dune_memory with
         | Some files ->
           let retrieve (file : Dune_memory.File.t) =
-            Log.infof "retrieve %s from cache"
-              (Path.to_string file.in_the_build_directory);
-            Unix.link
-              (Path.to_string file.in_the_memory)
-              (Path.to_string file.in_the_build_directory);
+            let path =
+              Filename.concat
+                (Path.to_string Path.build_dir)
+                (Path.to_string file.in_the_build_directory)
```

Comment:
This `concat` looks strange to me.
Doesn't `file.in_the_build_directory` already include the `_build` prefix? If not, then it shouldn't have the type `Path.t`. It could be `Path.Build.t` though (with a careful to/from string conversion).

Reply:
The issue is that this comes out of `dune_memory`, which reuses `Path.t` for path operations, but has no concept of build directory : it is standalone and can run outside of dune, in `dune-manager` for instance. Maybe I should stop using `Path.t` altogether in there and use `Filename` instead ?

Reply:
If the paths are always external, you can use `Path.External.t`. Otherwise feel free to add another path type to `path.ml`. I think that we should be using the Path module consistently in dune.

Reply:
`Path.Build.t` does not require knowledge of build directory. It's a path relative to an abstract idea of build directory, and I guess you still have that abstract idea even if not the directory itself?
If so, `Path.Build.t` is a good fit.